### DISCUSSION TITLE
How can pumgen efficiently convert msh format files to puml.h5 files?

### DISCUSSION THREAD

**Discussion by buzhiqiuyuebai** [SeisSol/SeisSol#878] [Q&A] [answered]
I am using pumgen with the following parameters on my computer, but when I use 4 or 8 mpirun to run it, it doesn't seem to work because the cpu utilization still adds up to 200%, which is a waste of time when doing large grid conversions. Is there any solution for this please?

![image](https://github.com/SeisSol/SeisSol/assets/114474894/b846047a-8e7e-439f-9b37-3ff41153663f)

This is `mpirun -np 4 pumgen prefix.msh -s msh2`

![image](https://github.com/SeisSol/SeisSol/assets/114474894/85d9b09b-da3d-4dec-8421-2f8f113e03f9)

This is `mpirun -np 8 pumgen prefix.msh -s msh2`

![image](https://github.com/SeisSol/SeisSol/assets/114474894/ea684d77-5496-4af0-85c0-37e73b8e4aec)

**Comment by Thomas-Ulrich** [comment, marked as answer]
I suppose the  Ascii msh file must be read sequentially, that's why it does not scale.

**Reply by buzhiqiuyuebai** [reply]
> I suppose the Ascii msh file must be read sequentially, that's why it does not scale.

Thank you for your kind suggestion!  By the way, how do you deal with the large-scale grid format conversion problem? Does the process take a long time?

**Reply by Thomas-Ulrich** [reply]
As far as I know, only @ravil-mobile has generated very large meshes with gmsh.
Typically we rely on SimModeler, which is embedded into pumgen, so there is no ASCII intermediate file.

**Reply by buzhiqiuyuebai** [reply]
I noticed that  gmsh can export a stl file and then [meshing with SimModeler](https://seissol.readthedocs.io/en/latest/gmsh.html#:~:text=may%20be%20obtained.-,gmsh%20to%20SimModeler,-%EF%83%81), I want to use this [file](https://github.com/buzhiqiuyuebai/GmshToSimModeler/blob/main/loh2.geo) to generate a 3D meshes,So I export 2D meshes with Binary and ASCII stl file,How can I use SimModeler to generate 3D meshes with stl file? How does Pumgen call SimModeler and use the stl format file to generate the mesh, do I need to install SimModeler separately or can I just use the command line to call SimModeler.(Generating .msh format files directly using gmsh is slow for large-scale grids, and there are sometimes problems in converting to puml.h5 format.)
When exporting a file in stl format, does the choice of Format and Solid have any effect on the subsequent generation of the mesh?
![image](https://github.com/SeisSol/SeisSol/assets/114474894/0ac9e554-2f2b-4db6-baaa-b96259667b10)
![image](https://github.com/SeisSol/SeisSol/assets/114474894/a7f7a780-3dc5-433a-90b0-54e133156261)
